File: src/main/java/cc/mirukuneko/realtrainmodrenewed/blockentity/SignalRemoteBlockEntity.java
package cc.mirukuneko.realtrainmodrenewed.blockentity;

import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewedBlockEntities;
import net.minecraft.core.BlockPos;
import net.minecraft.core.HolderLookup;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket;
import net.minecraft.world.level.storage.ValueInput;
import net.minecraft.world.level.storage.ValueOutput;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.state.BlockState;

/**
 * 受信機/変更機が共有する最小状態です。
 * 今はリンク済み信号番号だけ持てば十分なので、値を軽く保っています。
 */
public class SignalRemoteBlockEntity extends BlockEntity {
    private int linkedChannel = -1;

    public SignalRemoteBlockEntity(BlockPos pos, BlockState blockState) {
        super(RealTrainModRenewedBlockEntities.SIGNAL_REMOTE.get(), pos, blockState);
    }

    @Override
    protected void saveAdditional(ValueOutput tag) {
        super.saveAdditional(tag);
        tag.putInt("LinkedChannel", linkedChannel);
    }

    @Override
    protected void loadAdditional(ValueInput tag) {
        super.loadAdditional(tag);
        linkedChannel = tag.getIntOr("LinkedChannel", -1);
    }

    @Override
    public CompoundTag getUpdateTag(HolderLookup.Provider registries) {
        return saveWithoutMetadata(registries);
    }

    @Override
    public ClientboundBlockEntityDataPacket getUpdatePacket() {
        return ClientboundBlockEntityDataPacket.create(this);
    }

    public int getLinkedChannel() {
        return linkedChannel;
    }

    public void setLinkedChannel(int linkedChannel) {
        this.linkedChannel = linkedChannel;
        setChanged();
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/blockentity/SignalRemoteBlockEntity.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package cc.mirukuneko.realtrainmodrenewed.blockentity

import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewedBlockEntities
import net.minecraft.core.BlockPos
import net.minecraft.core.HolderLookup
import net.minecraft.nbt.CompoundTag
import net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket
import net.minecraft.world.level.storage.ValueInput
import net.minecraft.world.level.storage.ValueOutput
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.state.BlockState

class SignalRemoteBlockEntity(pos: BlockPos, state: BlockState) :
    BlockEntity(RealTrainModRenewedBlockEntities.SIGNAL_REMOTE.get(), pos, state) {

    @JvmField var linkedChannel: Int = -1

    fun getLinkedChannel(): Int = linkedChannel
    fun setLinkedChannel(channel: Int) { linkedChannel = channel; setChanged() }

    override fun saveAdditional(tag: ValueOutput) { super.saveAdditional(tag); tag.putInt("LinkedChannel", linkedChannel) }
    override fun loadAdditional(tag: ValueInput) { super.loadAdditional(tag); linkedChannel = tag.getIntOr("LinkedChannel", -1) }
    override fun getUpdateTag(registries: HolderLookup.Provider): CompoundTag = saveWithoutMetadata(registries)
    override fun getUpdatePacket(): ClientboundBlockEntityDataPacket = ClientboundBlockEntityDataPacket.create(this)
}
